core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -2204,6 +2207,25 @@ protected void inferUnknownTypes(
     }
   }
 
+  private boolean inferRowSqlNodeList(final RelDataType inferredType,
+      final SqlValidatorScope scope, final SqlNodeList nodeList) {
+    if (!inferredType.isStruct()) {
+      return false;
+    }
+    boolean inferred = false;
+    for (SqlNode child : nodeList) {
+      if (child instanceof SqlBasicCall && SqlKind.ROW == child.getKind()) {
```

Comment:
There are expressions that evaluate to a ROW type that are not ROW expressions.
For example, a[1], where a is an array of ROW values.